pkg/noderesourcetopology/score.go

```
@@ -0,0 +1,117 @@
+// it will return the minimal score of all the calculated NUMA's score, in order to avoid edge cases.
+func scoreForEachNUMANode(requested v1.ResourceList, numaList NUMANodeList, score scoreStrategy, resourceToWeightMap resourceToWeightMap) (int64, *framework.Status) {
+	numaScores := make([]int64, len(numaList))
+	for _, numa := range numaList {
+		numaScores[numa.NUMAID] = score(requested, numa.Resources, resourceToWeightMap)
+	}
+	klog.V(5).Infof("numaScores: %v", numaScores)
+	minScore := findMinScore(numaScores)
+	klog.V(5).Infof("node score: %v", minScore)
+	return minScore, nil
+}
+
+func calculateResourceAllocatableRequest(pod *v1.Pod, nodeTopology *topologyv1alpha1.NodeResourceTopology) (v1.ResourceList, NUMANodeList) {
+	containers := append(pod.Spec.InitContainers, pod.Spec.Containers...)
+	resources := make(v1.ResourceList)
+
+	for _, container := range containers {
+		for resource, quantity := range container.Resources.Requests {
+			if quan, ok := resources[resource]; ok {
+				quantity.Add(quan)
+			}
+			resources[resource] = quantity
+		}
+	}
+
+	allocatable := createNUMANodeList(nodeTopology.Zones)
+	return resources, allocatable
+}
+
+func findMinScore(arr []int64) int64 {
+	min := arr[0]
+
+	for _, score := range arr {
+		// if NUMA's score is 0, i.e. not fit at all, it won't be take under consideration by Kubelet.
+		if (min == 0) || (score != 0 && score < min) {
+			min = score
+		}
+	}
+	return min
+}
+
+func getScoringStrategyFunction(strategy apiconfig.ScoringStrategyType) scoreStrategy {
+	switch strategy {
+	case apiconfig.LeastAllocatable:
+		return leastAllocatableScoreStrategy
+	case apiconfig.MostAllocatable:
+		return mostAllocatableScoreStrategy
+	case apiconfig.BalancedAllocation:
+		return balancedAllocationScoreStrategy
+	// TBD what strategy should be provided by default
+	default:
+		return leastAllocatableScoreStrategy
```

The strategy should have defaulted earlier, so let's throw an error here.